rtnetlink/src/ns.rs

```diff
@@ -0,0 +1,223 @@
+                mount_flags.insert(nix::mount::MsFlags::MS_SHARED);
+                mount_flags.insert(nix::mount::MsFlags::MS_REC);
+                if nix::mount::mount(
+                    Some(Path::new("")),
+                    dir_path,
+                    Some(none_fs),
+                    mount_flags,
+                    none_p4,
+                )
+                .is_err()
+                {
+                    exit(-1);
+                }
+
+                let ns_path = Path::new(&netns_path);
+
+                // creating the netns file
+                let fd = match nix::fcntl::open(ns_path, open_flags, Mode::empty()) {
+                    Ok(raw_fd) => raw_fd,
+                    Err(_) => exit(-1),
+                };
+
+                if nix::unistd::close(fd).is_err() {
+                    let _ = nix::unistd::unlink(ns_path);
+                    exit(-1)
+                }
+
+                // unshare to the new network namespace
+                if nix::sched::unshare(CloneFlags::CLONE_NEWNET).is_err() {
+                    let _ = nix::unistd::unlink(ns_path);
+                    exit(-1);
+                }
+                open_flags = OFlag::empty();
+                open_flags.insert(OFlag::O_RDONLY);
+                open_flags.insert(OFlag::O_CLOEXEC);
+
+                let fd = match nix::fcntl::open(Path::new(&SELF_NS_PATH), open_flags, Mode::empty())
+                {
+                    Ok(raw_fd) => raw_fd,
+                    Err(_) => exit(-1),
+                };
+
+                let self_path = Path::new(&SELF_NS_PATH);
+
+                // bind to the netns
+                if nix::mount::mount(
+                    Some(self_path),
+                    ns_path,
+                    Some(none_fs),
+                    nix::mount::MsFlags::MS_BIND,
+                    none_p4,
+                )
+                .is_err()
+                {
+                    let _ = nix::unistd::unlink(ns_path);
+                    exit(-1);
+                }
+
+                setns_flags.insert(CloneFlags::CLONE_NEWNET);
+                if nix::sched::setns(fd, setns_flags).is_err() {
+                    let _ = nix::unistd::unlink(ns_path);
+                    exit(-1);
+                }
+
+                exit(0)
+            }
+            Err(e) => {
+                let err_msg = format!("Fork failed: {}", e);
+                Err(Error::NamespaceError(err_msg))
+            }
+        }
+    }
+
+    /// Remove a network namespace
+    /// This is equivalent to `ip netns del NS_NAME`.
+    pub async fn del(ns_name: String) -> Result<(), Error> {
+        let jh = task::spawn_blocking(move || {
+            let mut netns_path = String::new();
+            netns_path.push_str(NETNS_PATH);
+            netns_path.push_str(&ns_name);
+            let ns_path = Path::new(&netns_path);
+
+            if nix::mount::umount2(ns_path, nix::mount::MntFlags::MNT_DETACH).is_err() {
```

Comment:
Just to make sure I understand this correctly:

- `umount2` makes `/var/run/<namespace>` a mere directory, instead of a mount point
- `unlink` tells the system to delete `/var/run/<namespace>` after the last process that runs in that namespace exits.

Reply:
From: https://linux.die.net/man/2/umount2
I see that:
Linux 2.1.116 added the umount2() system call, which, like umount(), unmounts a target, but allows additional flags controlling the behavior of the operation: 
And `MNT_DETACH`
MNT_DETACH (since Linux 2.4.11)
    Perform a lazy unmount: make the mount point unavailable for new accesses, and actually perform the unmount when the mount point ceases to be busy. 

Those are the same calls done by `ip netns del <ns>`, see this portion of strace:
```
...
umount2("/run/netns/ns1", MNT_DETACH)   = 0
unlink("/run/netns/ns1")                = 0
...
```

Like `ip netns del <ns>` this means that if a process is still running in the namespace the namespace is actually removed when the last process in the namespace exits, but no new process can be started on the namespace.